compiler/rustc_expand/src/base.rs

```diff
@@ -761,6 +763,75 @@ impl SyntaxExtension {
         }
     }
 
+    fn collapse_debuginfo_by_name(sess: &Session, attr: &Attribute) -> CollapseMacroDebuginfo {
+        use crate::errors::CollapseMacroDebuginfoIllegal;
+        // #[collapse_debuginfo] without enum value (#[collapse_debuginfo(no/external/yes)])
+        // considered as `yes`
+        attr.meta_item_list().map_or(CollapseMacroDebuginfo::Yes, |l| {
+            let [NestedMetaItem::MetaItem(_item)] = &l[..] else {
+                sess.dcx().emit_err(CollapseMacroDebuginfoIllegal { span: attr.span });
+                return CollapseMacroDebuginfo::Unspecified;
+            };
+            let item = l[0].meta_item();
+            item.map_or(CollapseMacroDebuginfo::Unspecified, |mi| {
+                if !mi.is_word() {
+                    sess.dcx().emit_err(CollapseMacroDebuginfoIllegal { span: mi.span });
+                    CollapseMacroDebuginfo::Unspecified
+                } else {
+                    match mi.name_or_empty() {
+                        sym::no => CollapseMacroDebuginfo::No,
+                        sym::external => CollapseMacroDebuginfo::External,
+                        sym::yes => CollapseMacroDebuginfo::Yes,
+                        _ => {
+                            sess.dcx().emit_err(CollapseMacroDebuginfoIllegal { span: mi.span });
+                            CollapseMacroDebuginfo::Unspecified
+                        }
+                    }
+                }
+            })
+        })
+    }
+
+    /// if-ext - if macro from different crate (related to callsite code)
+    /// | cmd \ attr    | no  | (unspecified) | external | yes |
+    /// | no            | no  | no            | no       | no  |
+    /// | (unspecified) | no  | no            | if-ext   | yes |
+    /// | external      | no  | if-ext        | if-ext   | yes |
+    /// | yes           | yes | yes           | yes      | yes |
+    pub fn should_collapse(
```

Comment:
```suggestion
    fn should_collapse(
```

`get_collapse_debuginfo` and `should_collapse` can also be merged into a single function.